File: play-with-lambda-expressions/play_with_lambda/adapters/memory_names.py
"""In-memory named term registry with standard library preloaded."""

from __future__ import annotations

from ..domain import Term, parse, ParseError


STANDARD_LIBRARY: dict[str, str] = {
    "I": r"\x.x",
    "K": r"\x.\y.x",
    "S": r"\x.\y.\z.((x z) (y z))",
    "Y": r"\f.(\x.(f (x x)) \x.(f (x x)))",
    "OMEGA": r"(\x.(x x) \x.(x x))",
    "TRUE": r"\x.\y.x",
    "FALSE": r"\x.\y.y",
    "AND": r"\p.\q.((p q) p)",
    "OR": r"\p.\q.((p p) q)",
    "NOT": r"\p.((p \x.\y.y) \x.\y.x)",
    "ZERO": r"\f.\x.x",
    "SUCC": r"\n.\f.\x.(f ((n f) x))",
    "PLUS": r"\m.\n.\f.\x.((m f) ((n f) x))",
    "MULT": r"\m.\n.\f.(m (n f))",
}


class InMemoryNameRegistry:
    """Dict-backed name registry, preloaded with standard combinators.

    Implements the NameRegistry protocol.
    """

    def __init__(self) -> None:
        self._names: dict[str, Term] = {}
        for name, source in STANDARD_LIBRARY.items():
            result = parse(source)
            if isinstance(result, ParseError):
                continue
            self._names[name] = result

    def bind(self, name: str, term: Term) -> None:
        """Associate a name with a term (overwrites if exists).

        Args:
            name: The name to bind.
            term: The term to associate.
        """
        self._names[name] = term

    def lookup(self, name: str) -> Term | None:
        """Return the term bound to name, or None.

        Args:
            name: The name to look up.

        Returns:
            The bound term, or None if not found.
        """
        return self._names.get(name)

    def all_names(self) -> dict[str, Term]:
        """Return all name-term bindings.

        Returns:
            A shallow copy of the internal name registry.
        """
        return dict(self._names)

File: play-with-lambda-expressions/src/lambda_calculus/src/adapters/memory_names.rs
//! In-memory named term registry with standard library preloaded.

use std::collections::HashMap;
use crate::domain::{Term, parse};
use crate::ports::NameRegistry;

const STANDARD_LIBRARY: &[(&str, &str)] = &[
    ("I", r"\x.x"),
    ("K", r"\x.\y.x"),
    ("S", r"\x.\y.\z.((x z) (y z))"),
    ("Y", r"\f.(\x.(f (x x)) \x.(f (x x)))"),
    ("OMEGA", r"(\x.(x x) \x.(x x))"),
    ("TRUE", r"\x.\y.x"),
    ("FALSE", r"\x.\y.y"),
    ("AND", r"\p.\q.((p q) p)"),
    ("OR", r"\p.\q.((p p) q)"),
    ("NOT", r"\p.((p \x.\y.y) \x.\y.x)"),
    ("ZERO", r"\f.\x.x"),
    ("SUCC", r"\n.\f.\x.(f ((n f) x))"),
    ("PLUS", r"\m.\n.\f.\x.((m f) ((n f) x))"),
    ("MULT", r"\m.\n.\f.(m (n f))"),
];

/// Dict-backed name registry, preloaded with standard combinators.
pub struct InMemoryNameRegistry {
    names: HashMap<String, Term>,
}

impl InMemoryNameRegistry {
    pub fn new() -> Self {
        let mut names = HashMap::new();
        for &(name, source) in STANDARD_LIBRARY {
            if let Ok(term) = parse(source, None) {
                names.insert(name.to_string(), term);
            }
        }
        Self { names }
    }
}

impl NameRegistry for InMemoryNameRegistry {
    fn bind(&mut self, name: &str, term: Term) {
        self.names.insert(name.to_string(), term);
    }

    fn lookup(&self, name: &str) -> Option<&Term> {
        self.names.get(name)
    }

    fn all_names(&self) -> HashMap<String, Term> {
        self.names.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn standard_library_loaded() {
        let reg = InMemoryNameRegistry::new();
        assert_eq!(reg.all_names().len(), 14);
    }

    #[test]
    fn lookup_identity() {
        let reg = InMemoryNameRegistry::new();
        let i = reg.lookup("I").unwrap();
        assert_eq!(i.to_string(), "λx.x");
    }

    #[test]
    fn bind_and_lookup() {
        let mut reg = InMemoryNameRegistry::new();
        reg.bind("FOO", Term::var("bar"));
        assert_eq!(reg.lookup("FOO").unwrap(), &Term::var("bar"));
    }

    #[test]
    fn overwrite() {
        let mut reg = InMemoryNameRegistry::new();
        reg.bind("I", Term::var("overwritten"));
        assert_eq!(reg.lookup("I").unwrap(), &Term::var("overwritten"));
    }

    #[test]
    fn lookup_missing() {
        let reg = InMemoryNameRegistry::new();
        assert!(reg.lookup("NONEXISTENT").is_none());
    }

    #[test]
    fn all_names_is_copy() {
        let mut reg = InMemoryNameRegistry::new();
        let snapshot = reg.all_names();
        reg.bind("NEW", Term::var("x"));
        assert!(!snapshot.contains_key("NEW"));
    }
}
